uniffi_aries_vcx/core/src/handlers/holder.rs

```diff
@@ -0,0 +1,251 @@
+    OfferReceived,
+    RequestSet,
+    Finished,
+    Failed,
+}
+
+impl From<VcxHolderState> for HolderState {
+    fn from(x: VcxHolderState) -> Self {
+        match x {
+            VcxHolderState::Initial => HolderState::Initial,
+            VcxHolderState::ProposalSet => HolderState::ProposalSet,
+            VcxHolderState::OfferReceived => HolderState::OfferReceived,
+            VcxHolderState::RequestSet => HolderState::RequestSet,
+            VcxHolderState::Finished => HolderState::Finished,
+            VcxHolderState::Failed => HolderState::Failed,
+        }
+    }
+}
+
+pub fn create(source_id: String) -> VcxUniFFIResult<Arc<Holder>> {
+    block_on(async {
+        let handler = Mutex::new(VcxHolder::create(&source_id)?);
+        Ok(Arc::new(Holder { handler }))
+    })
+}
+
+pub fn create_from_offer(source_id: String, offer_message: String) -> VcxUniFFIResult<Arc<Holder>> {
+    let offer_message = serde_json::from_str(&offer_message)?;
+    block_on(async {
+        let handler = Mutex::new(VcxHolder::create_from_offer(&source_id, offer_message)?);
+        Ok(Arc::new(Holder { handler }))
+    })
+}
+
+pub fn create_with_proposal(
+    source_id: String,
+    propose_credential: String,
+) -> VcxUniFFIResult<Arc<Holder>> {
+    let propose_credential = serde_json::from_str(&propose_credential)?;
+
+    block_on(async {
+        let handler = Mutex::new(VcxHolder::create_with_proposal(
+            &source_id,
+            propose_credential,
+        )?);
+        Ok(Arc::new(Holder { handler }))
+    })
+}
+
+impl Holder {
+    pub fn set_proposal(&self, credential_proposal: String) -> VcxUniFFIResult<()> {
+        let mut handler = self.handler.lock()?;
+
+        handler.set_proposal(serde_json::from_str(&credential_proposal)?)?;
+        Ok(())
+    }
+
+    pub fn prepare_credential_request(
+        &self,
+        profile: Arc<ProfileHolder>,
+        my_pw_did: String,
+    ) -> VcxUniFFIResult<()> {
+        let mut handler = self.handler.lock()?;
+        let mut holder = handler.clone();
+
+        block_on(async {
+            holder
+                .prepare_credential_request(
+                    profile.inner.wallet(),
+                    profile.inner.ledger_read(),
+                    profile.inner.anoncreds(),
+                    my_pw_did,
+                )
+                .await?;
+            *handler = holder;
+
+            Ok(())
+        })
+    }
+
+    pub fn get_msg_credential_request(&self) -> VcxUniFFIResult<String> {
+        let handler = self.handler.lock()?;
+
+        Ok(serde_json::to_string(
+            &handler.clone().get_msg_credential_request()?,
+        )?)
+    }
+
+    pub fn decline_offer(&self, comment: Option<String>) -> VcxUniFFIResult<String> {
+        let handler = self.handler.lock()?;
+
+        Ok(serde_json::to_string(
+            &handler
+                .clone()
+                .decline_offer(Some(&comment.unwrap_or("".to_owned())))?,
+        )?)
```

Comment:
hmmm, whoops, i only just noticed this: i believe `decline_offer` is mutating the handler. so cloning it would result in that mutation not being registered.

also instead of `Some(&comment.unwrap_or("".to_owned())))` perhaps just `comment.as_deref()`